odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2743 @@
+
+        >>> z = r23.element([[1, 2, 3],
+        ...                  [4, 5, 6]])
+        >>> z.__array_ufunc__(np.add, 'reduce', z, axis=1)
+        rn(2, impl='cupy').element([  6.,  15.])
+
+        Finally, ``add.reduceat`` is a combination of ``reduce`` and
+        ``at`` with rather flexible and complex semantics (see the
+        `reduceat documentation`_ for details):
+
+        >>> x = r3.element([1, 2, 3])
+        >>> x.__array_ufunc__(np.add, 'reduceat', x, [0, 1])
+        rn(2, impl='cupy').element([ 1.,  5.])
+
+        References
+        ----------
+        .. _corresponding NEP:
+           https://github.com/numpy/numpy/blob/master/doc/neps/\
+ufunc-overrides.rst
+
+        .. _interface documentation:
+           https://github.com/charris/numpy/blob/master/doc/source/reference/\
+arrays.classes.rst#special-attributes-and-methods
+
+        .. _general documentation on Numpy ufuncs:
+           https://docs.scipy.org/doc/numpy/reference/ufuncs.html
+
+        .. _reduceat documentation:
+           https://docs.scipy.org/doc/numpy/reference/generated/\
+numpy.ufunc.reduceat.html
+        """
+        # --- Process `out` --- #
+
+        # Unwrap out if provided. The output parameters are all wrapped
+        # in one tuple, even if there is only one.
+        out_tuple = kwargs.pop('out', ())
+
+        # Check number of `out` args, depending on `method`
+        if method == '__call__' and len(out_tuple) not in (0, ufunc.nout):
+            raise ValueError(
+                "need 0 or {} `out` arguments for `method='__call__'`, "
+                'got {}'.format(ufunc.nout, len(out_tuple)))
+        elif method != '__call__' and len(out_tuple) not in (0, 1):
+            raise ValueError(
+                "need 0 or 1 `out` arguments for `method={!r}`, "
+                'got {}'.format(method, len(out_tuple)))
+
+        # We allow our own tensors, the data container type and
+        # `numpy.ndarray` objects as `out` (see docs for reason for the
+        # latter)
+        valid_types = (type(self), type(self.data), np.ndarray)
+        if not all(isinstance(o, valid_types) or o is None
+                   for o in out_tuple):
+            return NotImplemented
+
+        # Determine native ufunc vs. Numpy ufunc
+        if any(isinstance(o, np.ndarray) for o in out_tuple):
+            native_ufunc = None
+            use_native = False
+        else:
+            native_ufunc = getattr(cupy, ufunc.__name__, None)
+            # Manual assignment for sum, cumsum, prod and cumprod
+            if (ufunc in (np.add, np.multiply) and
+                    method in ('reduce', 'accumulate')):
+                use_native = native_ufunc is not None
+            else:
+                use_native = (native_ufunc is not None and
+                              hasattr(native_ufunc, method))
+
+        force_native = kwargs.pop('force_native', False)
+        if force_native and not use_native:
+            raise ValueError(
```

Comment:
I added this to give the user a way to ensure that no implicit dropdown to Numpy happens. You can do
```python
>>> rn = odl.rn((3, 4), impl='cupy')
>>> x = rn.one()
>>> np.add.reduce(x, force_native=True)
rn(4, impl='cupy').element([ 3.,  3.,  3.,  3.])
>>> np.divide.reduce(x, force_native=True)
Traceback (most recent call last):
  ...
ValueError: no native function available to evaluate `true_divide.reduce` [...]
```
It's really useful for debugging. Any objections?

Reply:
Functionality is great, but we may want to follow naming from other places, e.g. numba uses `nopython`:

http://numba.pydata.org/numba-doc/dev/user/jit.html

So here i guess `nofallback`?

Reply:
Well yeah, I kinda dislike the negative flags, in particular when the default is `False` like here. It's just unnecessarily hard to make out what it means. Compare `nofallback=False` with `force_native=False`. IMO the second is much clearer.
I'll look around for some other comparable flag name, `nofallback` is really not good.

Reply:
Maybe `allow_fallback=True` would be an alternative?